src/charm.py

```diff
@@ -2,21 +2,271 @@
+        """Generates UDR config file based on a given template and pushes it to the workload."""
+        content = self._render_config_file(
+            udr_ip_address=str(self._get_pod_ip()),
+            udr_sbi_port=UDR_SBI_PORT,
+            default_database_name=DEFAULT_DATABASE_NAME,
+            default_database_url=self._get_database_data()["uris"].split(",")[0],
+            nrf_url=self._nrf.nrf_url,
+        )
+        if not self._config_file_content_matches(content=content):
+            self._push_udr_config_file_to_workload(content=content)
+            self._configure_udr_service(force_restart=True)
+
+    def _configure_udr_service(self, force_restart: bool = False) -> None:
+        """Manages UDR's pebble layer and service.
+
+        Updates the pebble layer if the proposed config is different from the current one. If layer
+        has been updated also restart the workload service.
+
+        Args:
+            force_restart (bool): Allows for forcibly restarting the service even if Pebble plan
+                didn't change.
+        """
+        pebble_layer = self._pebble_layer
+        plan = self._container.get_plan()
+        if plan.services != pebble_layer.services or force_restart:
+            self._container.add_layer(self._container_name, pebble_layer, combine=True)
+            self._container.restart(self._service_name)
+            logger.info(f"Restarted container {self._service_name}")
+
+    @staticmethod
+    def _render_config_file(
+        *,
+        udr_ip_address: str,
+        udr_sbi_port: int,
+        default_database_name: str,
+        default_database_url: str,
+        nrf_url: str,
+    ) -> str:
+        """Renders the config file content.
+
+        Args:
+            udr_ip_address (str): UDR IP address.
+            udr_sbi_port (str): UDR SBI port.
+            default_database_name (str): Database name.
+            default_database_url (str): Database URL.
+            nrf_url (str): NRF URL.
+
+        Returns:
+            str: Config file content.
+        """
+        jinja2_env = Environment(loader=FileSystemLoader("src/templates"))
+        template = jinja2_env.get_template("udrcfg.yaml.j2")
+        return template.render(
+            udr_ip_address=udr_ip_address,
+            udr_sbi_port=udr_sbi_port,
+            default_database_name=default_database_name,
+            default_database_url=default_database_url,
+            nrf_url=nrf_url,
+        )
+
+    def _config_file_content_matches(self, content: str) -> bool:
+        """Returns whether the config file content matches the provided content.
+
+        Returns:
+            bool: Whether the config file content matches
+        """
+        udr_config_file = f"{BASE_CONFIG_PATH}/{UDR_CONFIG_FILE_NAME}"
+        try:
+            existing_content = self._container.pull(path=udr_config_file)
+            return existing_content.read() == content
+        except PathError:
+            return False
```

Comment:
If there's an exception reading the file wouldn't it raise `PathError`?
```suggestion
        except PathError:
            return False
        return existing_content.read() == content
```

Reply:
I don't think read() raises PathError.

Reply:
Of course not, my bad.